File: rekall-mcp/src/main/java/dev/rekall/mcp/protocol/JsonRpc.java
package dev.rekall.mcp.protocol;

import com.fasterxml.jackson.annotation.JsonInclude;
import tools.jackson.databind.JsonNode;

import java.util.Map;

public final class JsonRpc {

    public static final String VERSION = "2.0";

    public static final int PARSE_ERROR = -32700;

    public static final int INVALID_REQUEST = -32600;
    public static final int METHOD_NOT_FOUND = -32601;
    public static final int INVALID_PARAMS = -32602;
    public static final int INTERNAL_ERROR = -32603;

    public static final int HEADER_MISMATCH = -32020;

    public static final int UNSUPPORTED_PROTOCOL_VERSION = -32022;

    private JsonRpc() {
    }

    @JsonInclude(JsonInclude.Include.NON_NULL)
    public record Request(String jsonrpc, JsonNode id, String method, JsonNode params) {

        public boolean isNotification() {
            return id == null || id.isNull();
        }
    }

    @JsonInclude(JsonInclude.Include.NON_NULL)
    public record Response(String jsonrpc, JsonNode id, Object result, Error error) {

        public static Response success(JsonNode id, Object result) {
            return new Response(VERSION, id, result, null);
        }

        public static Response failure(JsonNode id, int code, String message) {
            return new Response(VERSION, id, null, new Error(code, message, null));
        }

        public static Response failure(JsonNode id, int code, String message, Map<String, Object> data) {
            return new Response(VERSION, id, null, new Error(code, message, data));
        }
    }

    @JsonInclude(JsonInclude.Include.NON_NULL)
    public record Error(int code, String message, Map<String, Object> data) {
    }
}

File: rekall-mcp/src/protocol/json_rpc.rs
use serde::Serialize;
use serde_json::{Map, Value};

pub const VERSION: &str = "2.0";

pub const PARSE_ERROR: i32 = -32700;
pub const INVALID_REQUEST: i32 = -32600;
pub const METHOD_NOT_FOUND: i32 = -32601;
pub const INVALID_PARAMS: i32 = -32602;
pub const INTERNAL_ERROR: i32 = -32603;
pub const HEADER_MISMATCH: i32 = -32020;
pub const UNSUPPORTED_PROTOCOL_VERSION: i32 = -32022;

/// A request as Jackson bound it: every field optional, `id` and `params` any JSON.
#[derive(Clone, Debug, Default)]
pub struct Request {
    pub jsonrpc: Option<String>,
    pub id: Option<Value>,
    pub method: Option<String>,
    pub params: Option<Value>,
}

impl Request {
    /// Reads the fields the way Jackson bound `JsonRpc.Request`: a JSON `null` is an absent field,
    /// and a scalar where a string belongs is coerced to its text.
    pub fn from_value(value: &Value) -> Self {
        let field = |name: &str| value.get(name).filter(|v| !v.is_null());
        Self {
            jsonrpc: field("jsonrpc").map(as_text),
            id: field("id").cloned(),
            method: field("method").map(as_text),
            params: field("params").cloned(),
        }
    }

    pub fn is_notification(&self) -> bool {
        self.id.is_none()
    }

    /// The id to answer with: the request's own, or JSON null.
    pub fn response_id(&self) -> Value {
        self.id.clone().unwrap_or(Value::Null)
    }
}

/// Jackson's `asString()`: a string as itself, a number or boolean as its text, anything else empty.
pub fn as_text(value: &Value) -> String {
    match value {
        Value::String(s) => s.clone(),
        Value::Number(n) => n.to_string(),
        Value::Bool(b) => b.to_string(),
        _ => String::new(),
    }
}

/// `node.hasNonNull(field) ? node.get(field).asString() : null`.
pub fn text_at(node: Option<&Value>, field: &str) -> Option<String> {
    node.and_then(|n| n.get(field)).filter(|v| !v.is_null()).map(as_text)
}

#[derive(Clone, Debug, Serialize)]
pub struct RpcError {
    pub code: i32,
    pub message: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub data: Option<Map<String, Value>>,
}

/// `@JsonInclude(NON_NULL)`: a success carries `result`, a failure `error`; `id` is always there.
#[derive(Clone, Debug, Serialize)]
pub struct Response {
    pub jsonrpc: &'static str,
    pub id: Value,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub result: Option<Value>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error: Option<RpcError>,
}

impl Response {
    pub fn success(id: Value, result: Value) -> Self {
        Self { jsonrpc: VERSION, id, result: Some(result), error: None }
    }

    pub fn failure(id: Value, code: i32, message: impl Into<String>) -> Self {
        Self { jsonrpc: VERSION, id, result: None, error: Some(RpcError { code, message: message.into(), data: None }) }
    }

    pub fn failure_with(id: Value, code: i32, message: impl Into<String>, data: Map<String, Value>) -> Self {
        Self { jsonrpc: VERSION, id, result: None, error: Some(RpcError { code, message: message.into(), data: Some(data) }) }
    }
}
